File: src/heretek_swarm/api/configuration.py
"""
Configuration API Endpoints

HTTP endpoints for managing system configurations, LLM providers, and embedding providers.
Provides CRUD operations through a RESTful API.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query

import structlog

from heretek_swarm.config.service import (
    ConfigurationService,
    get_config_service,
)
from heretek_swarm.config.models import (
    UserConfigurationCreate,
    UserConfigurationUpdate,
    LLMProviderCreate,
    LLMProviderUpdate,
    LLMProviderTestRequest,
    EmbeddingProviderCreate,
    EmbeddingProviderUpdate,
    EmbeddingProviderTestRequest,
    AgentConfigCreate,
    AgentConfigUpdate,
    ConfigurationImport,
    ImportOptions,
)
from heretek_swarm.gateway.auth import verify_auth

# Use lazy imports to break circular dependency at module load time
# These imports are resolved only when the functions are called
from heretek_swarm.utils import get_lazy_import

# Lazy-loaded provider factory functions
_llm_factory = get_lazy_import('heretek_swarm.llm.providers.factory')
_embedding_factory = get_lazy_import('heretek_swarm.embeddings.providers.factory')


def _get_llm_provider_factory():
    """Get LLM provider factory functions with lazy resolution."""
    return _llm_factory


def _get_embedding_provider_factory():
    """Get embedding provider factory functions with lazy resolution."""
    return _embedding_factory

logger = structlog.get_logger("api.configuration")

router = APIRouter(prefix="/api/config", tags=["Configuration"])


# =============================================================================
# Helper Functions
# =============================================================================

def get_service() -> ConfigurationService:
    """Dependency injection for ConfigurationService."""
    return get_config_service()


# =============================================================================
# User Configuration Endpoints
# =============================================================================

@router.get("")
async def get_all_configs(
    category: Optional[str] = Query(None, description="Filter by category"),
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Get all configurations with optional filtering."""
    configs = await service.list_configs(category=category, limit=limit, offset=offset)
    return {
        "configurations": [c.model_dump() for c in configs],
        "total": len(configs),
        "limit": limit,
        "offset": offset,
    }


@router.get("/{key}")
async def get_config(
    key: str,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Get a specific configuration by key."""
    config = await service.get_config(key)
    if not config:
        raise HTTPException(404, f"Configuration '{key}' not found")
    return config.model_dump()


@router.put("/{key}")
async def update_config(
    key: str,
    update: UserConfigurationUpdate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Update a configuration."""
    config = await service.update_config(key, update, changed_by=authenticated)
    if not config:
        raise HTTPException(404, f"Configuration '{key}' not found")
    return config.model_dump()


@router.post("")
async def create_config(
    config: UserConfigurationCreate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Create a new configuration."""
    try:
        new_config = await service.create_config(config, changed_by=authenticated)
        return new_config.model_dump()
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.delete("/{key}")
async def delete_config(
    key: str,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Delete a configuration."""
    success = await service.delete_config(key, changed_by=authenticated)
    if not success:
        raise HTTPException(404, f"Configuration '{key}' not found")
    return {"status": "deleted", "key": key}


# =============================================================================
# LLM Provider Endpoints
# =============================================================================

@router.get("/llm/types")
async def list_llm_provider_types() -> Dict[str, Any]:
    """List available LLM provider types."""
    types = list_llm_provider_types()
    info = [get_llm_provider_info(t) for t in types]
    return {"provider_types": info}


@router.get("/llm/providers")
async def list_llm_providers(
    provider_type: Optional[str] = Query(None, description="Filter by provider type"),
    enabled_only: bool = Query(False, description="Only return enabled providers"),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """List configured LLM providers."""
    providers = await service.list_llm_providers(
        provider_type=provider_type,
        enabled_only=enabled_only,
    )
    return {
        "providers": [p.model_dump() for p in providers],
        "total": len(providers),
    }


@router.get("/llm/providers/{provider_id}")
async def get_llm_provider(
    provider_id: UUID,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Get a specific LLM provider."""
    provider = await service.get_llm_provider(provider_id)
    if not provider:
        raise HTTPException(404, f"LLM provider '{provider_id}' not found")
    return provider.model_dump()


@router.post("/llm/providers")
async def create_llm_provider(
    provider: LLMProviderCreate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Create a new LLM provider."""
    try:
        # Note: API key should be encrypted before storage in production
        new_provider = await service.create_llm_provider(provider, changed_by=authenticated)
        return new_provider.model_dump()
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.put("/llm/providers/{provider_id}")
async def update_llm_provider(
    provider_id: UUID,
    update: LLMProviderUpdate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Update an LLM provider."""
    provider = await service.update_llm_provider(provider_id, update, changed_by=authenticated)
    if not provider:
        raise HTTPException(404, f"LLM provider '{provider_id}' not found")
    return provider.model_dump()


@router.delete("/llm/providers/{provider_id}")
async def delete_llm_provider(
    provider_id: UUID,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Delete an LLM provider."""
    success = await service.delete_llm_provider(provider_id, changed_by=authenticated)
    if not success:
        raise HTTPException(404, f"LLM provider '{provider_id}' not found")
    return {"status": "deleted", "id": str(provider_id)}


@router.post("/llm/providers/{provider_id}/test")
async def test_llm_provider(
    provider_id: UUID,
    test_request: LLMProviderTestRequest,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Test LLM provider connectivity."""
    provider = await service.get_llm_provider(provider_id)
    if not provider:
        raise HTTPException(404, f"LLM provider '{provider_id}' not found")
    
    try:
        # Create provider instance
        llm_provider = _get_llm_provider_factory().create_llm_provider(
            provider.provider_type,
            {
                "base_url": provider.base_url,
                "api_key": "test-key",  # In production, decrypt the actual key
                "default_model": test_request.model or provider.default_model,
                "extra_config": provider.extra_config,
            }
        )
        
        # Test connectivity
        result = await llm_provider.test_connectivity(model=test_request.model)
        
        return {
            "provider_id": str(provider_id),
            "provider_name": provider.provider_name,
            "success": result.success,
            "model_used": result.model_used,
            "latency_ms": result.latency_ms,
            "response_text": result.response_text if result.success else None,
            "error": result.error if not result.success else None,
        }
        
    except Exception as e:
        return {
            "provider_id": str(provider_id),
            "provider_name": provider.provider_name,
            "success": False,
            "error": str(e),
        }


# =============================================================================
# Embedding Provider Endpoints
# =============================================================================

@router.get("/embedding/types")
async def list_embedding_provider_types() -> Dict[str, Any]:
    """List available embedding provider types."""
    types = list_embedding_provider_types()
    info = [get_embedding_provider_info(t) for t in types]
    return {"provider_types": info}


@router.get("/embedding/providers")
async def list_embedding_providers(
    provider_type: Optional[str] = Query(None, description="Filter by provider type"),
    enabled_only: bool = Query(False, description="Only return enabled providers"),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """List configured embedding providers."""
    providers = await service.list_embedding_providers(
        provider_type=provider_type,
        enabled_only=enabled_only,
    )
    return {
        "providers": [p.model_dump() for p in providers],
        "total": len(providers),
    }


@router.get("/embedding/providers/{provider_id}")
async def get_embedding_provider(
    provider_id: UUID,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Get a specific embedding provider."""
    provider = await service.get_embedding_provider(provider_id)
    if not provider:
        raise HTTPException(404, f"Embedding provider '{provider_id}' not found")
    return provider.model_dump()


@router.post("/embedding/providers")
async def create_embedding_provider(
    provider: EmbeddingProviderCreate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Create a new embedding provider."""
    try:
        new_provider = await service.create_embedding_provider(provider, changed_by=authenticated)
        return new_provider.model_dump()
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.put("/embedding/providers/{provider_id}")
async def update_embedding_provider(
    provider_id: UUID,
    update: EmbeddingProviderUpdate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Update an embedding provider."""
    provider = await service.update_embedding_provider(provider_id, update, changed_by=authenticated)
    if not provider:
        raise HTTPException(404, f"Embedding provider '{provider_id}' not found")
    return provider.model_dump()


@router.delete("/embedding/providers/{provider_id}")
async def delete_embedding_provider(
    provider_id: UUID,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Delete an embedding provider."""
    success = await service.delete_embedding_provider(provider_id, changed_by=authenticated)
    if not success:
        raise HTTPException(404, f"Embedding provider '{provider_id}' not found")
    return {"status": "deleted", "id": str(provider_id)}


@router.post("/embedding/providers/{provider_id}/test")
async def test_embedding_provider(
    provider_id: UUID,
    test_request: EmbeddingProviderTestRequest,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Test embedding provider connectivity."""
    provider = await service.get_embedding_provider(provider_id)
    if not provider:
        raise HTTPException(404, f"Embedding provider '{provider_id}' not found")
    
    try:
        # Create provider instance
        embedding_provider = _get_embedding_provider_factory().create_embedding_provider(
            provider.provider_type,
            {
                "base_url": provider.base_url,
                "api_key": "test-key",  # In production, decrypt the actual key
                "default_model": test_request.model or provider.default_model,
                "extra_config": provider.extra_config,
            }
        )
        
        # Test connectivity by generating an embedding
        import time
        start_time = time.time()
        response = await embedding_provider.embed(
            texts=[test_request.text],
            model=test_request.model,
        )
        latency_ms = (time.time() - start_time) * 1000
        
        return {
            "provider_id": str(provider_id),
            "provider_name": provider.provider_name,
            "success": True,
            "model_used": response.model,
            "dimensions": response.dimensions,
            "latency_ms": latency_ms,
        }
        
    except Exception as e:
        return {
            "provider_id": str(provider_id),
            "provider_name": provider.provider_name,
            "success": False,
            "error": str(e),
        }


# =============================================================================
# Agent Configuration Endpoints
# =============================================================================

@router.get("/agent/configs")
async def list_agent_configs(
    agent_type: Optional[str] = Query(None, description="Filter by agent type"),
    active_only: bool = Query(True, description="Only return active configs"),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """List agent configurations."""
    configs = await service.list_agent_configs(agent_type=agent_type, active_only=active_only)
    return {
        "configs": [c.model_dump() for c in configs],
        "total": len(configs),
    }


@router.get("/agent/configs/{config_id}")
async def get_agent_config(
    config_id: UUID,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Get a specific agent configuration."""
    config = await service.get_agent_config(config_id)
    if not config:
        raise HTTPException(404, f"Agent config '{config_id}' not found")
    return config.model_dump()


@router.post("/agent/configs")
async def create_agent_config(
    config: AgentConfigCreate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Create a new agent configuration."""
    try:
        new_config = await service.create_agent_config(config, changed_by=authenticated)
        return new_config.model_dump()
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.put("/agent/configs/{config_id}")
async def update_agent_config(
    config_id: UUID,
    update: AgentConfigUpdate,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Update an agent configuration."""
    config = await service.update_agent_config(config_id, update, changed_by=authenticated)
    if not config:
        raise HTTPException(404, f"Agent config '{config_id}' not found")
    return config.model_dump()


@router.delete("/agent/configs/{config_id}")
async def delete_agent_config(
    config_id: UUID,
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Delete an agent configuration."""
    success = await service.delete_agent_config(config_id, changed_by=authenticated)
    if not success:
        raise HTTPException(404, f"Agent config '{config_id}' not found")
    return {"status": "deleted", "id": str(config_id)}


# =============================================================================
# Audit Log Endpoints
# =============================================================================

@router.get("/audit-log")
async def get_audit_log(
    entity_type: Optional[str] = Query(None, description="Filter by entity type"),
    limit: int = Query(100, ge=1, le=1000),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Get configuration audit log."""
    logs = await service.get_audit_log(entity_type=entity_type, limit=limit)
    return {
        "logs": [log.model_dump() for log in logs],
        "total": len(logs),
    }


# =============================================================================
# Import/Export Endpoints
# =============================================================================

@router.get("/export")
async def export_configurations(
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Export all configurations."""
    export_data = await service.export_configurations(exported_by=authenticated)
    return export_data.model_dump()


@router.post("/import")
async def import_configurations(
    import_data: ConfigurationImport,
    options: ImportOptions = ImportOptions(),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Import configurations from a bundle."""
    result = await service.import_configurations(import_data, options, changed_by=authenticated)
    return result.model_dump()


# =============================================================================
# Migration Endpoint
# =============================================================================

@router.post("/migrate-from-env")
async def migrate_from_env(
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """Migrate configurations from environment variables to database."""
    result = await service.migrate_from_env(changed_by=authenticated)
    return result


# =============================================================================
# Configuration Reload Endpoint
# =============================================================================

@router.post("/reload")
async def reload_configurations(
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """
    Reload configuration cache from database.
    
    This endpoint invalidates the cached configurations and reloads them
    from the database. Useful for applying configuration changes at runtime
    without restarting the application.
    
    Returns:
        Reload status and cache statistics
    """
    from heretek_swarm.config.loader import get_config_loader, reload_config
    
    try:
        # Reload the configuration cache
        reload_result = await reload_config()
        
        # Get cache stats
        loader = get_config_loader()
        cache_stats = loader.get_cache_stats()
        
        logger.info("Configuration reloaded", cached_keys=reload_result.get("cached_keys", []))
        
        return {
            "status": "reloaded",
            "cached_keys": reload_result.get("cached_keys", []),
            "cache_count": reload_result.get("cache_count", 0),
            "cache_stats": cache_stats,
            "reloaded_by": authenticated,
            "reloaded_at": datetime.utcnow().isoformat() if hasattr(datetime, 'utcnow') else datetime.now().isoformat(),
        }
    except Exception as e:
        logger.error("Configuration reload failed", error=str(e))
        raise HTTPException(500, f"Configuration reload failed: {e}")


# =============================================================================
# Configuration Health Check Endpoint
# =============================================================================

@router.get("/health")
async def configuration_health(
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """
    Check configuration service health status.
    
    Returns:
        Health status including database connectivity and cache status
    """
    from heretek_swarm.config.loader import get_config_loader
    
    try:
        # Test database connectivity
        test_config = await service.get_config("system.health_check")
        database_healthy = test_config is not None or True  # Config may not exist but connection works
        
        # Get cache stats
        loader = get_config_loader()
        cache_stats = loader.get_cache_stats()
        
        return {
            "status": "healthy",
            "database_connected": database_healthy,
            "cache_entries": cache_stats.get("total_entries", 0),
            "cache_hit_rate": cache_stats.get("hit_rate", 0),
        }
    except Exception as e:
        return {
            "status": "unhealthy",
            "database_connected": False,
            "error": str(e),
        }


# =============================================================================
# Configuration Import/Export Enhanced Endpoints
# =============================================================================

@router.get("/export/bundle")
async def export_configuration_bundle(
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """
    Export all configurations as a downloadable bundle.
    
    Returns:
        Complete configuration export with metadata
    """
    
    export_data = await service.export_configurations(exported_by=authenticated)
    
    return {
        "version": export_data.version,
        "exported_at": export_data.exported_at.isoformat(),
        "exported_by": export_data.exported_by,
        "user_configurations": [uc.model_dump() for uc in export_data.user_configurations],
        "llm_providers": [p.model_dump() for p in export_data.llm_providers],
        "embedding_providers": [p.model_dump() for p in export_data.embedding_providers],
        "agent_configs": [c.model_dump() for c in export_data.agent_configs],
        "summary": {
            "user_config_count": len(export_data.user_configurations),
            "llm_provider_count": len(export_data.llm_providers),
            "embedding_provider_count": len(export_data.embedding_providers),
            "agent_config_count": len(export_data.agent_configs),
        },
    }


@router.post("/import/bundle")
async def import_configuration_bundle(
    import_data: ConfigurationImport,
    options: ImportOptions = ImportOptions(),
    authenticated: str = Depends(verify_auth),
    service: ConfigurationService = Depends(get_service),
) -> Dict[str, Any]:
    """
    Import configurations from a bundle.
    
    Args:
        import_data: Configuration bundle to import
        options: Import options (skip_conflicts, etc.)
        authenticated: Authenticated user
        
    Returns:
        Import result summary
    """
    result = await service.import_configurations(import_data, options, changed_by=authenticated)
    
    return {
        "success": result.success,
        "imported_count": result.imported_count,
        "skipped_count": result.skipped_count,
        "error_count": result.error_count,
        "errors": result.errors,
        "imported_by": authenticated,
        "imported_at": datetime.utcnow().isoformat(),
    }

File: dashboard/frontend/src/api/configuration.ts
/**
 * Configuration API Client
 *
 * TypeScript client for interacting with the configuration API endpoints.
 * Provides type-safe methods for managing LLM providers, embedding providers,
 * and system configurations.
 */

import apiClient from './client';

// =============================================================================
// Types
// =============================================================================

export interface UserConfiguration {
  id: string;
  config_key: string;
  config_value: any;
  config_type: 'string' | 'integer' | 'float' | 'boolean' | 'json' | 'array';
  description?: string;
  category: string;
  is_sensitive: boolean;
  is_editable: boolean;
  created_at: string;
  updated_at: string;
}

export interface LLMProvider {
  id: string;
  provider_name: string;
  provider_type: 'openai' | 'openai_compatible' | 'ollama' | 'llamacpp' | 'zai' | 'minimax' | 'lemonade';
  base_url: string;
  api_key_hint?: string;
  default_model?: string;
  available_models: string[];
  is_enabled: boolean;
  is_default: boolean;
  priority: number;
  health_status: 'healthy' | 'unhealthy' | 'unknown' | 'degraded';
  created_at: string;
  updated_at: string;
}

export interface EmbeddingProvider {
  id: string;
  provider_name: string;
  provider_type: 'openai' | 'openai_compatible' | 'ollama' | 'local' | 'huggingface';
  base_url: string;
  api_key_hint?: string;
  default_model?: string;
  embedding_dimensions?: number;
  is_enabled: boolean;
  is_default: boolean;
  priority: number;
  health_status: 'healthy' | 'unhealthy' | 'unknown' | 'degraded';
  created_at: string;
  updated_at: string;
}

export interface AgentConfig {
  id: string;
  agent_type: string;
  agent_id?: string;
  config_name: string;
  config_data: Record<string, any>;
  llm_provider_id?: string;
  embedding_provider_id?: string;
  is_active: boolean;
  is_default_for_type: boolean;
  description?: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface LLMProviderCreate {
  provider_name: string;
  provider_type: string;
  base_url: string;
  api_key?: string;
  api_key_hint?: string;
  default_model?: string;
  is_enabled?: boolean;
  is_default?: boolean;
  priority?: number;
  extra_config?: Record<string, any>;
}

export interface EmbeddingProviderCreate {
  provider_name: string;
  provider_type: string;
  base_url: string;
  api_key?: string;
  api_key_hint?: string;
  default_model?: string;
  embedding_dimensions?: number;
  is_enabled?: boolean;
  is_default?: boolean;
  priority?: number;
}

export interface ProviderTestResult {
  success: boolean;
  provider_name: string;
  model_used: string;
  latency_ms: number;
  response_text?: string;
  error?: string;
}

export interface EmbeddingProviderTestResult {
  success: boolean;
  provider_name: string;
  model_used: string;
  dimensions?: number;
  latency_ms: number;
  error?: string;
}

// =============================================================================
// Configuration Endpoints
// =============================================================================

export const configurationApi = {
  // User Configurations
  getConfigs: async (category?: string): Promise<UserConfiguration[]> => {
    const params = category ? { category } : {};
    const response = await apiClient.get('/api/config', { params });
    return response.data.configurations;
  },

  getConfig: async (key: string): Promise<UserConfiguration> => {
    const response = await apiClient.get(`/api/config/${key}`);
    return response.data;
  },

  updateConfig: async (key: string, value: any): Promise<UserConfiguration> => {
    const response = await apiClient.put(`/api/config/${key}`, { config_value: value });
    return response.data;
  },

  createConfig: async (config: Partial<UserConfiguration>): Promise<UserConfiguration> => {
    const response = await apiClient.post('/api/config', config);
    return response.data;
  },

  deleteConfig: async (key: string): Promise<void> => {
    await apiClient.delete(`/api/config/${key}`);
  },

  // LLM Providers
  listLLMProviders: async (providerType?: string, enabledOnly?: boolean): Promise<LLMProvider[]> => {
    const params: Record<string, string> = {};
    if (providerType) params.provider_type = providerType;
    if (enabledOnly) params.enabled_only = 'true';
    const response = await apiClient.get('/api/config/llm/providers', { params });
    return response.data.providers;
  },

  getLLMProvider: async (id: string): Promise<LLMProvider> => {
    const response = await apiClient.get(`/api/config/llm/providers/${id}`);
    return response.data;
  },

  createLLMProvider: async (provider: LLMProviderCreate): Promise<LLMProvider> => {
    const response = await apiClient.post('/api/config/llm/providers', provider);
    return response.data;
  },

  updateLLMProvider: async (id: string, provider: Partial<LLMProviderCreate>): Promise<LLMProvider> => {
    const response = await apiClient.put(`/api/config/llm/providers/${id}`, provider);
    return response.data;
  },

  deleteLLMProvider: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/config/llm/providers/${id}`);
  },

  testLLMProvider: async (id: string, prompt?: string, model?: string): Promise<ProviderTestResult> => {
    const response = await apiClient.post(`/api/config/llm/providers/${id}/test`, {
      prompt: prompt || 'Hello, this is a connectivity test.',
      model,
      max_tokens: 10,
    });
    return response.data;
  },

  listLLMProviderTypes: async (): Promise<any[]> => {
    const response = await apiClient.get('/api/config/llm/types');
    return response.data.provider_types;
  },

  // Embedding Providers
  listEmbeddingProviders: async (providerType?: string, enabledOnly?: boolean): Promise<EmbeddingProvider[]> => {
    const params: Record<string, string> = {};
    if (providerType) params.provider_type = providerType;
    if (enabledOnly) params.enabled_only = 'true';
    const response = await apiClient.get('/api/config/embedding/providers', { params });
    return response.data.providers;
  },

  getEmbeddingProvider: async (id: string): Promise<EmbeddingProvider> => {
    const response = await apiClient.get(`/api/config/embedding/providers/${id}`);
    return response.data;
  },

  createEmbeddingProvider: async (provider: EmbeddingProviderCreate): Promise<EmbeddingProvider> => {
    const response = await apiClient.post('/api/config/embedding/providers', provider);
    return response.data;
  },

  updateEmbeddingProvider: async (id: string, provider: Partial<EmbeddingProviderCreate>): Promise<EmbeddingProvider> => {
    const response = await apiClient.put(`/api/config/embedding/providers/${id}`, provider);
    return response.data;
  },

  deleteEmbeddingProvider: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/config/embedding/providers/${id}`);
  },

  testEmbeddingProvider: async (id: string, text?: string, model?: string): Promise<EmbeddingProviderTestResult> => {
    const response = await apiClient.post(`/api/config/embedding/providers/${id}/test`, {
      text: text || 'This is a test sentence for embedding.',
      model,
    });
    return response.data;
  },

  listEmbeddingProviderTypes: async (): Promise<any[]> => {
    const response = await apiClient.get('/api/config/embedding/types');
    return response.data.provider_types;
  },

  // Agent Configurations
  listAgentConfigs: async (agentType?: string, activeOnly?: boolean): Promise<AgentConfig[]> => {
    const params: Record<string, string> = {};
    if (agentType) params.agent_type = agentType;
    if (activeOnly !== undefined) params.active_only = String(activeOnly);
    const response = await apiClient.get('/api/config/agent/configs', { params });
    return response.data.configs;
  },

  getAgentConfig: async (id: string): Promise<AgentConfig> => {
    const response = await apiClient.get(`/api/config/agent/configs/${id}`);
    return response.data;
  },

  createAgentConfig: async (config: Partial<AgentConfig>): Promise<AgentConfig> => {
    const response = await apiClient.post('/api/config/agent/configs', config);
    return response.data;
  },

  updateAgentConfig: async (id: string, config: Partial<AgentConfig>): Promise<AgentConfig> => {
    const response = await apiClient.put(`/api/config/agent/configs/${id}`, config);
    return response.data;
  },

  deleteAgentConfig: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/config/agent/configs/${id}`);
  },

  // Import/Export
  exportConfigurations: async (): Promise<any> => {
    const response = await apiClient.get('/api/config/export');
    return response.data;
  },

  importConfigurations: async (data: any, options?: any): Promise<any> => {
    const response = await apiClient.post('/api/config/import', {
      import_data: data,
      options: options || {},
    });
    return response.data;
  },

  // Migration
  migrateFromEnv: async (): Promise<any> => {
    const response = await apiClient.post('/api/config/migrate-from-env');
    return response.data;
  },
};

export default configurationApi;
